### PR TITLE
8302354: InstanceKlass init state/thread should be atomic

### PR DESCRIPTION
**Pull request by robehn**
Hi, please consider.

These states are read 'lock-less' should thus be atomic.
Also the assert setting init thread state can be stricter.

Passes t1-4

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302354](https://bugs.openjdk.org/browse/JDK-8302354): InstanceKlass init state/thread should be atomic

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [595c2349](https://git.openjdk.org/jdk/pull/12542/files/595c234967fcdfc1d3015ed86a2a0faa43868970)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [595c2349](https://git.openjdk.org/jdk/pull/12542/files/595c234967fcdfc1d3015ed86a2a0faa43868970)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12542/head:pull/12542` \
`$ git checkout pull/12542`

Update a local copy of the PR: \
`$ git checkout pull/12542` \
`$ git pull https://git.openjdk.org/jdk pull/12542/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12542`

View PR using the GUI difftool: \
`$ git pr show -t 12542`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12542.diff">https://git.openjdk.org/jdk/pull/12542.diff</a>

</details>

### CODE REVIEW

**Review by coleenp — APPROVED** [review]
This looks good but can you edit the bug and PR to not have so many 'sss' in InstanceKlass ?

**Comment by robehn** [comment]
> This looks good but can you edit the bug and PR to not have so many 'sss' in InstanceKlass ?

Thanks! Fixed!

File: src/hotspot/share/oops/instanceKlass.hpp [hunk 1]
```diff
@@ -226,15 +226,15 @@ class InstanceKlass: public Klass {
...
 
   u1              _reference_type;          // reference type
 
   // State is set while executing, eventually atomically to not disturb other state
   InstanceKlassFlags _misc_flags;
 
-  Monitor*        _init_monitor;         // mutual exclusion to _init_state and _init_thread.
-  Thread*         _init_thread;          // Pointer to current thread doing initialization (to handle recursive initialization)
+  Monitor*             _init_monitor;           // mutual exclusion to _init_state and _init_thread.
+  JavaThread* volatile _init_thread;            // Pointer to current thread doing initialization (to handle recursive initialization)
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/instanceKlass.hpp:237]: Nit: please re-align comments with preceding lines. Thanks

**Review by dholmes-ora — APPROVED** [review]
Scratch previous review comments. Use of Atomic::load/store guards against theoretical possibility of non-atomic ptr read/writes. I was confusing with visibility issues. Sorry for the noise.

**Comment by robehn** [comment]
> Scratch previous review comments. Use of Atomic::load/store guards against theoretical possibility of non-atomic ptr read/writes. I was confusing with visibility issues. Sorry for the noise.

Thanks! Fixed "re-align comments".

**Pull request closed without merging**